src/Castle.Windsor/MicroKernel/Arguments.cs

```diff
@@ -160,6 +125,111 @@ IEnumerator IEnumerable.GetEnumerator()
 			return GetEnumerator();
 		}
 
+		/// <summary>
+		///   Inserts a new named argument with given key. If an argument for this name already exists, it will be overwritten.
+		/// </summary>
+		public Arguments Insert(string key, object value) 
```

Comment:
What was the reason for moving these from extension methods to methods on `Arguments`? I feel they fit exactly to what extension methods should be used for. I do like the changes to update them to align with the other changes.

Reply:
Why? 

I think we need to discuss why you think implementing extension methods are valid when you have access to source code. If you own the source the co-related type you should really just implement it as a member on that type. 

Extensions methods are typically viable for types where you don't have access to the source. 

I am clearly missing something here.

Reply:
@jnm2 Do you think extensions are viable even when you have access to the source?

Reply:
I question this approach when we are not doing things at scale like ASP.NET Core.

Reply:
@fir3pho3nixx The .NET BCL team is big on extension methods instead of instance methods for generic types because each instance method adds to the memory and CPU required to JIT the type before it can be used at all. That doesn't apply here.

Beyond that, there's no problem per se with using extension methods as part of your API. You can be forced to due to assembly layering or wanting stricter generic constraints for the method than the type. This doesn't result in a problematic API, so I wouldn't thing of it as being problematic to use extension methods when you aren't forced to.

It's really up to the semantics you want to convey: is this a core use case, without which the type would not be properly usable? Instance feels right to me for that. Is it add-on behavior? Extension methods can express that, though I do default to instance members myself.

One practical difference:
Extension methods require a using statement to be in scope, but instance methods do not. This is a great tool because you can hide extension methods behind a namespace on purpose to avoid clutter for people who don't opt in.
On the flip side, I've been annoyed a few times at having to add `using System.Reflection;` because System.Reflection.TypeExtensions backfills APIs using extension methods, where other target frameworks have instance methods and no need for the `using`.

Reply:
> Why?
> 
> I think we need to discuss why you think implementing extension methods are valid when you have access to source code. If you own the source the co-related type you should really just implement it as a member on that type.
> 
> Extensions methods are typically viable for types where you don't have access to the source.
> 
> I am clearly missing something here.

The comment @jnm2 left is really where I thought is the reason:

> It's really up to the semantics you want to convey: is this a core use case, without which the type would not be properly usable? Instance feels right to me for that. Is it add-on behavior? Extension methods can express that, though I do default to instance members myself.

These Insert methods were extension methods before (e.g. in master) and you even looked at removing them because they aren't necessary, but found them too useful as helper methods because you can fluently insert because they return the collection. 

It just feels a bit strange to me that we've got 2 methods `Add(K,V)` and `Insert(K,V)`, insert only taking a string, they feel like helpers, unless you intended to make `Add(object, object)` explicitly implemented and hide it wanting people to use Insert rather than Add?

```csharp
public object this[object key]
public void Add(object key, object value)
public Arguments Insert(string key, object value)
public Arguments Insert(Type dependencyType, object value)
public Arguments InsertProperties(object instance)
public Arguments InsertTyped<TDependencyType>(TDependencyType value)
```

However, I'll leave the call to you since you feel strongly on this one, this is a minor thing in the scheme of things.

Reply:
Thanks @jnm2. Always very informative chatting to you ;)

I still breaking this into extensions is a bit too much. What are we defending against? What are we trying compartmentalise? At it’s core these are methods which help govern how kv’s are inserted into a dictionary. Not sure what building these as extensions gives us except that we have moved them somewhere else and hidden them behind a namespace one has to opt into. These methods are useful and used extensively throughout the tests. Unless we are saying we want to get rid of them in future I am not sure.

Reply:
> I still breaking this into extensions is a bit too much. What are we defending against? What are we trying compartmentalise? At it’s core these are methods which help govern how kv’s are inserted into a dictionary.

I guess at first it felt like the factory methods were the ones that knew what to do. Would you change the factory methods to call these since these are the ones that are meant to govern how names and types become keys, rather than both having the same implementation?

> Not sure what building these as extensions gives us except that we have moved them somewhere else and hidden them behind a namespace one has to opt into.

Never intended them to be hidden in an obscure namespace.

Please answer this question, I think this is my sticking point trying to understand what method a user is meant to use:

> It just feels a bit strange to me that we've got 2 methods `Add(K,V)` and `Insert(K,V)`, insert only taking a string, they feel like helpers, unless you intended to make `Add(object, object)` explicitly implemented and hide it wanting people to use Insert rather than Add?

Reply:
> I guess at first it felt like the factory methods were the ones that knew what to do. Would you change the factory methods to call these since these are the ones that are meant to govern how names and types become keys, rather than both having the same implementation?

Yes. I would take harmony and encapsulation over `not yet discoverable API's` until you read the docs any day. All day long. 

> Never intended them to be hidden in an obscure namespace.

Sorry man did not mean obscure. Hidden was my point.

> Please answer this question, I think this is my sticking point trying to understand what method a user is meant to use:

> It just feels a bit strange to me that we've got 2 methods Add(K,V) and Insert(K,V), insert only taking a string, they feel like helpers, unless you intended to make Add(object, object) explicitly implemented and hide it wanting people to use Insert rather than Add?

`Add` throws an ArgumentException if the key already exists, `Insert` overwrites. Fuzzy as to how this applies to implementing API's as extension methods. Can you explain a little further?

Reply:
If a user wanted to add an argument to the collection after creating it, which way would you direct them (either via documentation or if asked directly), for example::

```csharp
var args = new Arguments();

// [... other code here ...]

args.Add("connString", "Server=...");
// or
args.Insert("connString", "Server=...");
```

I was asking whether you'd always direct people to the Insert methods, but obviously Add stays there to implement `IDictionary`.

Reply:
There is no mutual exclusivity here. With the previous arrangement (where insert was an extension method) we had a situation where people would have to opt in via namespace/scope, so if I had to guess what we were saying there was this: "Please use `Add` first and leave them to discover `Insert` later".   

Without `Insert` as an extension what we are saying now is: "You can add or insert, up to you really". `Add` if you want a guarantee against the value being added already, `Insert` if you want to overwrite or take advantage of the fluent interface it offers eg. `args.Insert(...).Insert(...)`. The offering here is semantically different enough, such that both could co-exist. 

Now, if this is bad design(Add vs Insert) in your mind, that is something altogether different again. So what I am trying to get to the bottom of here is why they cannot co-exist in the same class as instances methods but implementing the `least preferred` method as an extension is OK. Stuff like this should be jumping up at you when you look at a class asking you to refactor it. You also end up with the namespaced fractal design around your API's which simply serves to hide complexity. Maybe what this needs is refinement to harmonise things?

Reply:
> Now, if this is bad design(Add vs Insert) in your mind, that is something altogether different again. So what I am trying to get to the bottom of here is why they cannot co-exist in the same class as instances methods but implementing the `least preferred` method as an extension is OK.

I'm happy for this to stay, you've convinced me it is useful enough to be instance methods. Let's just make the static factory methods also use these.

Reply:
Just to jokingly add more fuel to the fire, we could get rid of the static factory methods. :)

Serious I don't mind changing it. My only recommendation is that we probably raise a new issue, to track it and do it as a patch release after V5.

Reply:
> Just to jokingly add more fuel to the fire, we could get rid of the static factory methods. :)
>
> Serious I don't mind changing it if you agree.

Might be a good idea as it'll simplify things and mean there isn't 2 sets of APIs. What would the alternatives look like, there isn't one for the last 2:

```
Arguments.FromDictionary(dictionary)
new Arguments(dictionary)

Arguments.FromProperties(instance)
new Arguments().InsertProperties(instance)

Arguments.FromReadOnlyDictionary(readOnlyDictionary)
?

Arguments.FromTypedArguments(typedArguments)
?
```

Reply:
```
Arguments.FromDictionary(dictionary)
new Arguments(dictionary)

Arguments.FromProperties(instance)
new Arguments().InsertProperties(instance)

Arguments.FromReadOnlyDictionary(readOnlyDictionary)
new Arguments().InsertReadOnly(instance)

Arguments.FromTypedArguments(typedArguments)
new Arguments().InsertTypedArguments(instance)
```

Continue with the insert convention?

Reply:
> Probably should add Insert(IDictionary) to complete the set, and make it check the key is of type string or Type.

I have checked keys to be `string` or `object` for `Insert(IDictionary)` and to only be `string` for `InsertNamed(IReadOnlyDictionary<string, object>)`. 

> Should we go with InsertTyped rather than InsertTypedArguments?

Like it. Done. 

> You've not responded to my comment. 

Sorry missed that. I have renamed the method to `InsertNamed`. This works well as the `string` enforces that things are named only.